File: framework/src/datamirai_engine/tools/builtin/filesystem/edit_file.py
"""filesystem/edit_file — exact search-and-replace editing."""

from __future__ import annotations

import os
from typing import Any

from datamirai_engine.core.context import ExecutionContext
from datamirai_engine.tools.base import (
    BaseTool,
    ConfigField,
    ToolInput,
    ToolOutput,
    ToolSpec,
)


class EditFileTool(BaseTool):
    spec = ToolSpec(
        tool_type="filesystem/edit_file",
        version="1.0.0",
        display_name="Edit File",
        description="Performs exact string replacement in a file. The old_string must match exactly (including whitespace). By default replaces only the first occurrence.",
        category="filesystem",
        icon="file-edit",
        intents=[
            "editar un archivo reemplazando texto exacto",
            "modificar una parte especifica de un archivo",
            "hacer search-and-replace en un archivo",
        ],
        inputs=[
            ToolInput(name="path", type="string", required=True, description="Path to the file to edit"),
            ToolInput(name="old_string", type="string", required=True, description="Exact string to find"),
            ToolInput(name="new_string", type="string", required=True, description="Replacement string"),
        ],
        outputs=[
            ToolOutput(name="path", type="string", description="Resolved absolute path"),
            ToolOutput(name="replacements", type="number", description="Number of replacements made"),
            ToolOutput(name="diff", type="string", description="Summary of changes"),
        ],
        config=[
            ConfigField(name="replace_all", type="boolean", default=False, description="Replace all occurrences instead of just the first"),
        ],
    )

    async def execute(
        self,
        inputs: dict[str, Any],
        config: dict[str, Any],
        context: ExecutionContext | None,
    ) -> dict[str, Any]:
        path = os.path.abspath(inputs["path"])
        old_string = inputs["old_string"]
        new_string = inputs["new_string"]
        replace_all = config.get("replace_all", False)

        if not os.path.isfile(path):
            raise FileNotFoundError(f"File not found: {path}")

        with open(path, "r", encoding="utf-8", errors="replace") as f:
            content = f.read()

        if old_string == new_string:
            raise ValueError("old_string and new_string are identical")

        count = content.count(old_string)
        if count == 0:
            raise ValueError(f"old_string not found in {path}")

        if not replace_all and count > 1:
            raise ValueError(
                f"old_string found {count} times in {path}. "
                "Provide more context to make it unique, or set replace_all=true."
            )

        if replace_all:
            new_content = content.replace(old_string, new_string)
            replacements = count
        else:
            new_content = content.replace(old_string, new_string, 1)
            replacements = 1

        with open(path, "w", encoding="utf-8") as f:
            f.write(new_content)

        old_preview = old_string[:80].replace("\n", "\\n")
        new_preview = new_string[:80].replace("\n", "\\n")
        diff = f"-  {old_preview}\n+  {new_preview}\n({replacements} replacement(s))"

        return {
            "path": path,
            "replacements": replacements,
            "diff": diff,
        }

File: engine/src/tools/builtin/filesystem/edit_file.rs
//! EditFileTool

use super::*;

// ===========================================================================
// EditFileTool
// ===========================================================================

fs_tool! {
    struct EditFileTool, factory EditFileFactory;
    tool_type = "filesystem/edit_file",
    name = "Edit File",
    description = "Performs exact string replacement in a file. The old_string must match exactly. By default replaces only the first occurrence.",
    category = "filesystem",
    inputs = [
        field("path", FieldType::String, true, "Path to the file to edit"),
        field("old_string", FieldType::String, true, "Exact string to find"),
        field("new_string", FieldType::String, true, "Replacement string"),
    ],
    outputs = [
        field("path", FieldType::String, true, "Resolved absolute path"),
        field("replacements", FieldType::Number, true, "Number of replacements made"),
        field("diff", FieldType::String, true, "Summary of changes"),
    ],
    config_fields = [
        field("replace_all", FieldType::Boolean, false, "Replace all occurrences instead of just the first"),
    ]
}

#[async_trait]
impl Tool for EditFileTool {
    async fn execute(
        &self,
        inputs: HashMap<String, Value>,
        config: &HashMap<String, Value>,
        _context: &dyn ExecutionContext,
    ) -> Result<HashMap<String, Value>, ToolError> {
        let raw_path = inputs
            .get("path")
            .and_then(|v| v.as_str())
            .ok_or_else(|| ToolError::ExecutionFailed {
                tool_type: "filesystem/edit_file".into(),
                message: "missing required input: path".into(),
            })?;
        let old_string = inputs
            .get("old_string")
            .and_then(|v| v.as_str())
            .ok_or_else(|| ToolError::ExecutionFailed {
                tool_type: "filesystem/edit_file".into(),
                message: "missing required input: old_string".into(),
            })?;
        let new_string = inputs
            .get("new_string")
            .and_then(|v| v.as_str())
            .ok_or_else(|| ToolError::ExecutionFailed {
                tool_type: "filesystem/edit_file".into(),
                message: "missing required input: new_string".into(),
            })?;
        let replace_all = config
            .get("replace_all")
            .and_then(|v| v.as_bool())
            .unwrap_or(false);

        let path = fs::canonicalize(raw_path).map_err(|e| ToolError::ExecutionFailed {
            tool_type: "filesystem/edit_file".into(),
            message: format!("File not found: {raw_path} ({e})"),
        })?;

        if !path.is_file() {
            return Err(ToolError::ExecutionFailed {
                tool_type: "filesystem/edit_file".into(),
                message: format!("Not a file: {}", path.display()),
            });
        }

        let content = fs::read_to_string(&path).map_err(|e| ToolError::ExecutionFailed {
            tool_type: "filesystem/edit_file".into(),
            message: format!("Cannot read file: {e}"),
        })?;

        if old_string == new_string {
            return Err(ToolError::ExecutionFailed {
                tool_type: "filesystem/edit_file".into(),
                message: "old_string and new_string are identical".into(),
            });
        }

        let count = content.matches(old_string).count();
        if count == 0 {
            return Err(ToolError::ExecutionFailed {
                tool_type: "filesystem/edit_file".into(),
                message: format!("old_string not found in {}", path.display()),
            });
        }

        if !replace_all && count > 1 {
            return Err(ToolError::ExecutionFailed {
                tool_type: "filesystem/edit_file".into(),
                message: format!(
                    "old_string found {} times in {}. Provide more context to make it unique, or set replace_all=true.",
                    count,
                    path.display()
                ),
            });
        }

        let (new_content, replacements) = if replace_all {
            (content.replace(old_string, new_string), count)
        } else {
            (content.replacen(old_string, new_string, 1), 1)
        };

        fs::write(&path, &new_content).map_err(|e| ToolError::ExecutionFailed {
            tool_type: "filesystem/edit_file".into(),
            message: format!("Cannot write file: {e}"),
        })?;

        // Build diff preview (truncate long strings)
        let old_preview: String = old_string.chars().take(80).collect::<String>().replace('\n', "\\n");
        let new_preview: String = new_string.chars().take(80).collect::<String>().replace('\n', "\\n");
        let diff = format!(
            "-  {}\n+  {}\n({} replacement(s))",
            old_preview, new_preview, replacements
        );

        let mut out = HashMap::new();
        out.insert("path".to_string(), json!(path.display().to_string()));
        out.insert("replacements".to_string(), json!(replacements));
        out.insert("diff".to_string(), json!(diff));
        Ok(out)
    }
}

